dot/state/grandpa_changes.go

```diff
@@ -0,0 +1,310 @@
+	// of a slice ordered by the effective number and by announcing header number
+	idxToInsert := sort.Search(oc.Len(), func(i int) bool {
+		return (*oc)[i].effectiveNumber() >= pendingChange.effectiveNumber() &&
+			(*oc)[i].announcingHeader.Number >= pendingChange.announcingHeader.Number
+	})
+
+	*oc = append(*oc, pendingChange)
+	copy((*oc)[idxToInsert+1:], (*oc)[idxToInsert:])
+	(*oc)[idxToInsert] = pendingChange
+	return nil
+}
+
+// pruneChanges will remove changes whose are not descendant of the hash argument
+// this function updates the current state of the change tree
+func (oc *orderedPendingChanges) pruneChanges(hash common.Hash, isDescendantOf isDescendantOfFunc) error {
+	onBranchForcedChanges := make([]*pendingChange, 0, oc.Len())
+
+	for _, forcedChange := range *oc {
+		isDescendant, err := isDescendantOf(hash, forcedChange.announcingHeader.Hash())
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			onBranchForcedChanges = append(onBranchForcedChanges, forcedChange)
+		}
+	}
+
+	*oc = onBranchForcedChanges
+	return nil
+}
+
+type pendingChangeNode struct {
+	change *pendingChange
+	nodes  []*pendingChangeNode
+}
+
+func (c *pendingChangeNode) importNode(blockHash common.Hash, blockNumber uint, pendingChange *pendingChange,
+	isDescendantOf isDescendantOfFunc) (imported bool, err error) {
+	announcingHash := c.change.announcingHeader.Hash()
+
+	if blockHash.Equal(announcingHash) {
+		return false, errDuplicateHashes
+	}
+
+	if blockNumber <= c.change.announcingHeader.Number {
+		return false, nil
+	}
+
+	for _, childrenNodes := range c.nodes {
+		imported, err := childrenNodes.importNode(blockHash, blockNumber, pendingChange, isDescendantOf)
+		if err != nil {
+			return false, fmt.Errorf("cannot track node: %w", err)
+		}
+
+		if imported {
+			return true, nil
+		}
+	}
+
+	isDescendant, err := isDescendantOf(announcingHash, blockHash)
+	if err != nil {
+		return false, fmt.Errorf("cannot check ancestry: %w", err)
+	}
+
+	if !isDescendant {
+		return false, nil
+	}
+
+	pendingChangeNode := &pendingChangeNode{change: pendingChange, nodes: []*pendingChangeNode{}}
+	c.nodes = append(c.nodes, pendingChangeNode)
+	return true, nil
+}
+
+// changeTree keeps track of the changes per fork allowing
+// n forks in the same structure, this structure is intended
+// to be an acyclic directed graph where the change nodes are
+// placed by descendency order and number, you can ensure an
+// node ancestry using the `isDescendantOfFunc`
+type changeTree []*pendingChangeNode
+
+func (ct changeTree) Len() int { return len(ct) }
+func (ct *changeTree) importChange(pendingChange *pendingChange, isDescendantOf isDescendantOfFunc) error {
+	for _, root := range *ct {
+		imported, err := root.importNode(pendingChange.announcingHeader.Hash(),
+			pendingChange.announcingHeader.Number, pendingChange, isDescendantOf)
+
+		if err != nil {
+			return err
```

Comment:
```suggestion
		if err != nil {
			return fmt.Errorf("cannot import node: %w", err)
```

Reply:
This error is already wrapped by the caller and by the calling function, so adding this wrapper the error will look like this: `cannot import scheduled change: cannot import node: cannot import node: ..."`